docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,473 @@
+simply drop this engine entirely.
+
+Of particular interest is "commands" - these involve communicating with the
+engine regarding commands targetting it, and accepting commands to be send to
+other devices. Note that outgoing commands are likely to not originate from a sync,
+but instead from other actions, such as "restore bookmarks".
+
+However, because the only current requirement for commands is to wipe the
+store, and because you could anticipate "wipe" also being used when remotely
+disconnecting a device (eg, when a device is lost or stolen), our lives would
+probably be made much simpler by initially supporting only per-engine wipe
+commands.
+
+## Collections vs engines vs stores vs preferences
+
+For the purposes of the sync manager, we define:
+
+* An *engine* is the unit exposed to the user - an "engine" can be enabled
+  or disabled. There is a single set of canonical "engines" used across the
+  entire sync ecosystem - ie, desktop and mobile devices all need to agree
+  about what engines exist and what the identifier for an engine is.
+
+* A *store* is the code which actually syncs. This is largely an implementation
+  detail. There may be multiple stores per engine (eg, the "history" engine
+  may have "history" and "forms" stores.)
+
+* A *collection* is a unit of storage on a server. It's even more of an
+  implementation detail than a store. For example, you might imagine a future
+  where the "history" store uses multiple "collections" to help with containers.
+
+In practice, this means that the high-level component should only need to care
+about an *engine*. The low-level component will manage the mapping of engines
+to stores.
+
+## The declined list
+
+This document isn't going to outline the history of how "declined" is used, nor
+talk about how this might change in the future. For the purposes of the sync
+manager, we have the following hard requirements:
+
+* The low-level component needs to know what the currently declined set of
+  engines is for the purposes of re-populating `meta/global`.
+
+* The low-level component needs to know when the declined set needs to change
+  based on user input (ie, when the user opts in to or out of a particular
+  engine on this device)
+
+* The high-level component needs to be aware that the set of declined engines
+  may change on every sync (ie, when the user opts in to or out of a particular
+  engine on another device)
+
+A complication is that due to networks being unreliable, there's an inherent
+conflict between "what is the current state?" and "what state changes are
+requested?". For example, if the user changes the state of an engine while
+there's no network, then exits the app, how do we ensure the user's new state
+is updated next time the app starts? What if the user has since made a
+different state request on a different device? Is the state as last-known on
+this device considered canonical?
+
+To clarify, consider:
```

Comment:
_whispers_ Three-way merging. I'm joking a bit, but also...what if the sync manager persisted the list of declined engines, recorded local changes since the last sync, and merged them with the server's view once the network is back up? That does push some complexity for noting declined state changes up into the high-level manager, which could be an argument for having the sync manager manage its own storage.

As you say, though, it's an extreme edge case, and having the manager manage its own database is probably not worth the benefit of handling conflicting `declined` changes.

Reply:
> _whispers_ Three-way merging.

We *could* do that - but *should* we?

Storing the transition is simple because it is a bool state - so just storing `(name, new_state)` seems enough - `old_state` can be implied. And we already store declined state, so adding this probably doesn't blow out "state budget" and force us into our own state storage.

It's more about user intent. It's like logins, where we *could* reconcile "device-a changed the username but not the password" and "device-b changed the password but not the username". We could reconcile and end up with both "new" values, but it seems unlikely to reflect user intent.

Similarly here - it's not clear to me "device-a declined bookmarks but left history enabled" but "device-b declined history and left bookmarks enabled" is best served by merging these rather than just taking the latest.

But yeah - "not clear" is the keyword here, so I'm erring on the side of simplicity - both in implementation and explanation - but I'm totally open to the fact I'm underthinking this :)